From bfcf5541c0428de859efedb27e2ae3b9d42461aa Mon Sep 17 00:00:00 2001
From: Peter Harliman Liem <peter.harliman.liem@intel.com>
Date: Tue, 31 Mar 2020 11:44:07 +0800
Subject: [PATCH] PONRTSYS-6960 - Add support for die selection in spinand

This is to add support for SPI flash with multi-die (at
the moment only Winbond WN25M02GV is using this feature).
---
 drivers/mtd/ltq-spinand/ltq_spinand.c | 87 ++++++++++++++++++++++++++++++++---
 drivers/mtd/ltq-spinand/ltq_spinand.h |  6 +++
 2 files changed, 86 insertions(+), 7 deletions(-)

diff --git a/drivers/mtd/ltq-spinand/ltq_spinand.c b/drivers/mtd/ltq-spinand/ltq_spinand.c
index ab280ae50ed0..1443cce7ceb5 100644
--- a/drivers/mtd/ltq-spinand/ltq_spinand.c
+++ b/drivers/mtd/ltq-spinand/ltq_spinand.c
@@ -28,6 +28,7 @@ struct spinand_flash_dev {
 	struct nand_flash_dev nand;
 
 	/* extension to standard nand flash structure */
+	u32 num_of_die;
 };
 
 /*============================================================================*/
@@ -165,6 +166,7 @@ static struct spinand_flash_dev spinand_flash_ids[] = {
 		},
 	{ {"SPI NAND 2Gbit 3,3v WN25M02GV",
 		{ .id = {0xef, 0xab} }, 2048, 256, 0x20000, LP_OPTIONS, 2, 64},
+		.num_of_die = 2,
 		},
 };
 #if 0
@@ -351,9 +353,11 @@ static int spi_nand_manufacture_init(struct mtd_info *mtd, struct nand_chip *chi
  */
 struct nand_flash_dev * spinand_flash_detect(struct mtd_info *mtd, struct nand_chip *chip)
 {
+	struct spinand_info *info = (struct spinand_info *)chip->priv;
 	u8 id_data[8];
 	int i;
 	struct nand_flash_dev *type;
+	u32 diesize;
 
 	chip->cmdfunc(mtd, NAND_CMD_READID, 0x00, -1);
 	for (i = 0; i < 3; i++)
@@ -387,6 +391,17 @@ struct nand_flash_dev * spinand_flash_detect(struct mtd_info *mtd, struct nand_c
 	/* do we need it ? */
 	chip->options |= type->options;
 
+	info->num_of_die = spinand_flash_ids[i].num_of_die;
+	if (!info->num_of_die)
+		info->num_of_die = 1; /* default to 1 if undefined */
+
+	/* fill up die information */
+	diesize = chip->chipsize >> (ffs(info->num_of_die) - 1);
+	info->page_per_die_shift = ffs(diesize) - ffs(mtd->writesize);
+	dev_dbg(&info->spi->dev,
+		"diesize 0x%x page_per_die %d\n",
+		diesize, info->page_per_die_shift);
+
 	/* manufacturer init */
 	spi_nand_manufacture_init(mtd, chip, type);
 
@@ -556,6 +571,44 @@ static int wait_till_ready(struct spi_device *spi)
 	return -1;
 }
 
+/**
+ * spinand_select_die - Select a specific die
+ * @spi: the spi device
+ * @die: die to select
+ *
+ * Select a new target/die.
+ */
+static int spinand_select_die(struct spi_device *spi, u32 die)
+{
+	struct spinand_info *info = spi_to_info(spi);
+	struct spinand_cmd cmd = {0};
+	int ret;
+	u8 buf;
+
+	if (info->current_die == die)
+		return 0;
+
+	if (die >= info->num_of_die) {
+		dev_err(&spi->dev, "Invalid die %d / %d\n",
+			die, info->num_of_die);
+		return -EINVAL;
+	}
+
+	cmd.cmd = CMD_SWITCH_DIE;
+	cmd.n_tx = 1;
+	cmd.tx_buf = &buf;
+	buf = die & 0xff;
+
+	ret = spinand_cmd(spi, &cmd);
+	if (ret < 0) {
+		dev_err(&spi->dev, "Failed to select die %d\n", die);
+		return ret;
+	}
+
+	info->current_die = die;
+	return 0;
+}
+
 /**
  * spinand_get_otp - [INTERN] send command 0xf to read the SPI Nand OTP register
  * @spi: the spi device.
@@ -793,9 +846,14 @@ static void spinand_ecc_status(struct spi_device *spi,
 static int spinand_read_page(struct spi_device *spi, int page_id,
 		u16 offset, u16 len, u8 *rbuf)
 {
+	struct spinand_info *info = spi_to_info(spi);
 	int ret, ecc_error;
 	u8 status = 0;
 
+	ret = spinand_select_die(spi, page_id >> info->page_per_die_shift);
+	if (ret)
+		return ret;
+
 	ret = spinand_read_page_to_cache(spi, page_id);
 	if (ret < 0)
 		return ret;
@@ -916,12 +974,16 @@ static int spinand_program_execute(struct spi_device *spi, int page_id)
 static int spinand_program_page(struct spi_device *spi,
 		int page_id, u16 offset, u16 len, u8 *buf)
 {
-	// struct mtd_info *mtd = (struct mtd_info *)dev_get_drvdata((const struct device *)&(spi->dev));
+	struct spinand_info *info = spi_to_info(spi);
 	int retval;
 	u8 status = 0;
 	uint8_t *wbuf;
 	int retry = 3;
 
+	retval = spinand_select_die(spi, page_id >> info->page_per_die_shift);
+	if (retval)
+		return retval;
+
 	wbuf = buf;
 
 	retval = spinand_write_enable(spi);
@@ -1012,9 +1074,14 @@ static int spinand_erase_block_erase(struct spi_device *spi, u32 block_id)
  */
 static int spinand_erase_block(struct spi_device *spi, u32 block_id)
 {
+	struct spinand_info *info = spi_to_info(spi);
 	int retval;
 	u8 status = 0;
 
+	retval = spinand_select_die(spi, block_id >> info->page_per_die_shift);
+	if (retval)
+		return retval;
+
 	retval = spinand_write_enable(spi);
 	if (wait_till_ready(spi))
 		dev_err(&spi->dev, "WAIT timedout!!!\n");
@@ -1080,7 +1147,9 @@ static int spinand_lock_block(struct spi_device *spi_nand, u8 lock)
  */
 static void spinand_reset(struct spi_device *spi)
 {
+	struct spinand_info *info = spi_to_info(spi);
 	struct spinand_cmd cmd = {0};
+	int i;
 
 	cmd.cmd = CMD_RESET;
 
@@ -1093,14 +1162,18 @@ static void spinand_reset(struct spi_device *spi)
 	if (wait_till_ready(spi))
 		dev_err(&spi->dev, "WAIT timedout!\n");
 
-	/* safety try to unlock after reset */
-	spinand_lock_block(spi, BL_ALL_UNLOCKED);
+	for (i = 0; i < info->num_of_die; i++) {
+		spinand_select_die(spi, i);
 
-	if (wait_till_ready(spi))
-		dev_err(&spi->dev, "WAIT timedout!\n");
+		/* safety try to unlock after reset */
+		spinand_lock_block(spi, BL_ALL_UNLOCKED);
 
-	/* safety enable On-Die ECC again in case */
-	spinand_set_otp_mask(spi, OTP_ECC_ENABLE, OTP_ECC_ENABLE);
+		if (wait_till_ready(spi))
+			dev_err(&spi->dev, "WAIT timedout!\n");
+
+		/* safety enable On-Die ECC again in case */
+		spinand_set_otp_mask(spi, OTP_ECC_ENABLE, OTP_ECC_ENABLE);
+	}
 
 	return;
 }
diff --git a/drivers/mtd/ltq-spinand/ltq_spinand.h b/drivers/mtd/ltq-spinand/ltq_spinand.h
index 9366086b25c1..6f756a39502a 100644
--- a/drivers/mtd/ltq-spinand/ltq_spinand.h
+++ b/drivers/mtd/ltq-spinand/ltq_spinand.h
@@ -41,6 +41,7 @@
 #define CMD_WRITE_REG                   0x1f
 #define CMD_READ_CACHE_X2               0x3b
 #define CMD_READ_CACHE_X4               0x6b
+#define CMD_SWITCH_DIE                  0xc2
 
 /* feature registers */
 #define REG_BLOCK_LOCK                  0xa0
@@ -90,6 +91,11 @@ struct spinand_info {
 	struct nand_ecclayout *ecclayout;
 	struct spi_device *spi;
 	void *priv;
+
+	/* for die selection */
+	u8 current_die;
+	u32 num_of_die;
+	u32 page_per_die_shift;
 };
 
 struct spinand_privbuf {
